src/inmanta/data/__init__.py

```diff
@@ -4815,6 +4815,293 @@ def get_compliance_status(self) -> Optional[state.Compliance]:
+            WHERE rscm.environment=$1
+                AND rscm.model=$2
+            GROUP BY r.resource_id
+            HAVING COUNT(*) > 1
+        """
+        records = await cls._fetch_query(query, environment, version, connection=connection)
+        if records:
+            rid_to_resource_sets: dict[str, dict[str, str]] = {}
+            for record in records:
+                resource_id = str(record["resource_id"])
+                resource_set_names = list(record["name"])
+                if len(resource_set_names) != 2:
+                    raise BadRequest(
+                        f"Resource {resource_id} appears on version {version} in these resource sets: {resource_set_names}"
+                    )
+                rid_to_resource_sets[resource_id] = {}
+                for name in resource_set_names:
+                    key = "new" if name in updated_resource_sets else "old"
+                    if key in rid_to_resource_sets[resource_id]:
+                        raise BadRequest(
+                            f"Resource set with name {name} appears more than once on version {version} "
+                            f"of the model with ids {rid_to_resource_sets[resource_id]} and {resource_id}"
+                        )
+                    rid_to_resource_sets[resource_id][key] = name
+
+            msg = (
+                "The following Resource(s) cannot be migrated to a different resource set using a partial compile, "
+                "a full compile is necessary for this process:\n"
+            )
+            msg += "\n".join(
+                f"    {rid} moved from {cls.get_printable_name_for_resource_set(resource_sets["old"])} "
+                f"to {cls.get_printable_name_for_resource_set(resource_sets["new"])}"
+                for rid, resource_sets in rid_to_resource_sets.items()
+            )
+            raise BadRequest(msg)
+
+    @classmethod
+    async def insert_sets_and_resources(
+        cls,
+        environment: uuid.UUID,
+        target_version: int,
+        updated_resources: list[m.Resource],
+        updated_resource_sets: list[str | None],
+        base_version: Optional[int] = None,
+        resource_set_names_not_to_bump: Optional[Set[str | None]] = None,
+        connection: Optional[asyncpg.connection.Connection] = None,
+    ) -> None:
+        """
+        Inserts resources and resource sets. These resource sets are linked to the target version
+        :param environment: The environment of these resources.
+        :param target_version: The version which we want to link the resource sets to
+        :param updated_resources: A list of resources to insert.
+            On a full compile, this list should contain all resources present in the version.
+            On a partial compile, this list should contain all resources belonging to a resource set that was changed.
+        :param updated_resource_sets: A list of resource sets that were changed.
```

Comment:
What is this for, if we also have `resource_set_names_not_to_bump` ?

Reply:
`resource_set_names_not_to_bump` includes updated_resource_sets and deleted_resource_sets